feat(spec): allow defining custom media type of responses + fix(koa): dont overwrite content-type of response

mrl5 opened the pull request:
problem description: currently `application/json` is hardcoded into generated spec
    
proposed solution: allow specifying custom media type in generated spec
on both controller level by:
  - using new `@Produces` decorator

and on method level by:
  - using new `@Produces` decorator
  - introducing new optional arg for `@SuccessResponse` decorator
  - introducing new optional arg for `@Response` decorator
  - setting `Content-Type` header in `@Res` decorator

impact:
  - cli: controller generator
  - cli: method generator
  - cli: parameter generator
  - cli: spec2 generator
  - cli: spec3 generator
  - runtime: response decorators

    
Closes #1126
Related #968

### All Submissions:

- [x] Have you followed the guidelines in our [Contributing](https://github.com/lukeautry/tsoa/tree/master/docs/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/lukeautry/tsoa/pulls) for the same update/change?
- [x] Have you written unit tests?
- [ ] Have you written unit tests that cover the negative cases (i.e.: if bad data is submitted, does the library respond properly)?
- [x] This PR is associated with an existing issue?

**Closing issues**

- [x] Put `closes #XXXX` (where XXXX is the issue number) in your comment to auto-close the issue that your PR fixes.

### If this is a new feature submission:

- [x] Has the issue had a maintainer respond to the issue and clarify that the feature is something that aligns with the [goals](https://github.com/lukeautry/tsoa#goal) and [philosophy](https://github.com/lukeautry/tsoa#philosophy) of the project?

**Potential Problems With The Approach**

<!-- if there are areas of the solution that you think might have limitations or potential "gotchas", then discuss them here -->
<!-- if you have many questions, then please consider discussing them in the issue thread (unless you need to share code to illustrate the questions) -->
1. Someone could think that defining custom media type will also override `content-type` of actual response (which is not a case). However I don't think it's a problem -> if you check the issues from the past you will see that writing custom methods was encouraged - for example #44 #224
2. OAS2 is not affected by this changes
3. ~~I wasn't able to introduce new arg for setting custom media type for `@Res` decorator~~


**Test plan**

<!-- explain why the unit tests that you added are demonstrating that the feature works -->
1. I've tried to show that it is possible to set `@Produces` on controller level
2. I've tried to show that even when `@Produces` is set on controller level it is possible to set it also for specific method
3. I've tried to show that OAS3 media type of responses is possible to be changed by passing new optional arg in `@Response` and `@SuccessResponse` decorators
4. I've tried to show that OAS3 media type of responses is possible to be changed by defining headers for `@Res` decorator

WoH commented on packages/runtime/src/decorators/response.ts:
```diff
@@ -23,6 +24,18 @@ export function Response<ExampleType, HeaderType extends IsValidHeader<HeaderTyp
  * The type of the responder function should be annotated `TsoaResponse<Status, Data, Headers>` in order to support OpenAPI documentation.
  */
 export function Res(): Function {
+  // todo: allow setting media type
```
I think it'd be better to check the Header type (3rd arg to `TsoaResponse`) over adding an arg to `@Res`.

dderevjanik commented:
@mrl5 very nice PR! 
I'm just curious, what's the reason behind `@MediaType` decorator? I mean, most frameworks (jersey, spring, asp.net, nestjs, etc.) are using `@Produces` or `@ProducesSomething` wording  - so you can distinguish between `@Consumes`. 

In case that TSOA will support `@Consumes` in future, IMHO it could be confusing to have something like this in controller:

```ts
 @Get('process')
 @Consumes('image/png')  // Consumes MediaType
 @MediaType('text/plain') // Produces MediaType
  public async processImage(@Request() request: ExRequest): Promise<string> {
    // code
    return '';
  }
```

mrl5 commented:
thanks for code review @dderevjanik - I agree that `@Produces` and `@Consumes` are more accurate. I will change that

> I'm just curious, what's the reason behind @MediaType decorator?
I was inspired by the code snippets from #968

@WoH is it ok to `git commit --amend` not to produce commit flood on every improvement or you prefer to keep it separated?

WoH commented:
Amending and rebasing is fine

mrl5 commented:
1. pushed improvements suggested by @WoH and @dderevjanik
2. updated PR description
3. added possibility to define `@Responses`` on controller level

mrl5 commented on tests/fixtures/controllers/mediaTypeController.ts:
```diff
@@ -0,0 +1,45 @@
+
+    const body = { id: model.name.length, name: model.name };
+    this.setStatus(202);
+    this.setHeader('content-type', 'application/vnd.mycompany.myapp.v2+json'); // express and hapi returns it, koa ignores
```
subject of #1134

mrl5 commented:
pushed additional commit that:
1. fixes issue #1134 
2. introduces integration tests checking `content-type`

WoH commented on packages/cli/src/swagger/specGenerator2.ts:
```diff
@@ -14,13 +14,13 @@ export class SpecGenerator2 extends SpecGenerator {
       },
       paths: this.buildPaths(),
-      produces: ['application/json'],
+      produces: [DEFAULT_RESPONSE_MEDIA_TYPE], // todo: Tsoa.Response.produces
```
> A list of MIME types the APIs can produce.
This is global to all APIs but can be overridden on specific API calls.
Value MUST be as described under Mime Types.

Maybe I'm interpreting too much, but we should be able to leave this as is, remove the todo and override on path.method.produces?

WoH commented on packages/cli/src/swagger/specGenerator2.ts:
```diff
@@ -222,7 +222,7 @@ export class SpecGenerator2 extends SpecGenerator {
     const operation: Swagger.Operation = {
       operationId: this.getOperationId(method.name),
-      produces: ['application/json'],
+      produces: [DEFAULT_RESPONSE_MEDIA_TYPE], // todo: Tsoa.Response.produces
```
We should fill in the `produces` here, right?

mrl5 replied:
> We should fill in the `produces` here, right?

as opposed to `packages/cli/src/swagger/specGenerator3.ts` where there are `buildPaths()` and `buildOperation()` this would require major refactor.

unfortunately I don't want to do it in this PR

mrl5 replied:
> We should fill in the `produces` here, right?

indeed, will do

mrl5 commented:
1. amended with changes from code review
2. updated PR description

thanks @WoH for detailed CR :) I hope we're good to go now

WoH commented on packages/cli/src/swagger/specGenerator2.ts:
```diff
@@ -220,9 +222,14 @@ export class SpecGenerator2 extends SpecGenerator {
       }
     });
 
+    produces = Array.from(new Set(produces.filter(p => p)));
```
`filter(p => p)` ?

WoH commented:
Thanks for this PR and your patience during the review, much appreciated. Great Work!

alberto-rubio commented:
Please, @mrl5  and @WoH, is this working in version 4.0.0-alpha.2? I'm adding the decorator:
`@Produces('text/event-stream')`
at method level, but it continues producing "application/json" in the generated openapi yaml definition.

Should I do any additional thing to see "text/event-stream" in the generate openapi?

Thanks and best regards!!

mrl5 commented:
hi @alberto-rubio - I think that this changes are already included in this version. It's hard to help you w/o seeing any example code that is producing unexpected results

if you need example you can study this file that is used in unit tests to ensure that feature introduced in this PR works: https://github.com/lukeautry/tsoa/blob/6a289a7ea0576835cf0803db9905a8f39cab0107/tests/fixtures/controllers/mediaTypeController.ts#L27-L28
